src/soundiness.rkt

```diff
@@ -37,14 +38,51 @@
                           num-decrease (length prev)))))
   proof-diffs)
   
+(define (generate-rewrite-once-proof rule loc prog prev)
+  (list (program-body (alt-program prev)) ;; Start Expression
+        (list 'Rewrite=> 
+              (rule-name rule) ;; name
+              (program-body prog)))) ;; expr
 
 (define (add-soundiness-to pcontext ctx altn)
   (match altn
-    [(alt prog `(simplify ,loc ,input #f #f) `(,prev))
-     (define proof (get-proof input
+    ;; This is alt coming from rr
+    [(alt prog `(rr, loc, input #f #f) `(,prev))
+      (cond
+        [(rr-input? input) ;; Check if input is an rr-input struct (B-E-R)
+          (define proof 
+            (get-rr-proof (rr-input-rules input)
+                          (rr-input-input-exprs input) (rr-input-iter-limit input)
+                          (location-get loc (alt-program prev)) (location-get loc prog)))
+          (cond
+            [proof
+              (define proof*
+                (for/list ([step proof])
+                  (let ([step* (canonicalize-rewrite step)])
+                    (program-body (location-do loc prog (λ _ step*))))))
+              (define errors
+                (let ([vars (program-variables prog)])
+                  (get-proof-errors proof* pcontext ctx vars)))
+              (alt prog `(rr, loc, input, proof* ,errors) `(,prev))]
+            [else
+              (alt prog `(rr ,loc, input #f #f) `(,prev))])]
+
+        [(rule? input) ;; (R-O) case
+          (define proof-ro
+            (generate-rewrite-once-proof input loc prog prev))
+          (define errors-ro
+            (let ([vars (program-variables prog)])
+              (get-proof-errors proof-ro pcontext ctx vars)))
+          (alt prog `(rr, loc, input, proof-ro, errors-ro) `(,prev))]
```

Comment:
This is kind of ugly, but I guess it's OK?

Reply:
We couldn't think of a better solution. @bksaiki and @JPepperwood123 found this was the best way.

Reply:
Yeah sorry about that. It's easier to hardcode the rewrite-once proof by matching rule as an input struct. Though patch.rkt still looks great and abstracts whether it is a rule or rr-input struct being passed.